### PR TITLE
Iar 25/andy/create a posts feed unit tests

### PR DESCRIPTION
**Pull request by AndyM22**
Completed unit tests.
Updated scss.
Updated folder structure + created new components to prepare for authentication.

### CODE REVIEW

File: src/app/app-routing.module.ts [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
 
 const routes: Routes = [
-    { path: '', redirectTo: 'feed', pathMatch: 'full' },
-    { path: 'feed', component: FeedComponent}
+    {
+        path: 'authorized',
+        component: AuthorizedComponent,
+        children: [{ path: 'feed', component: FeedComponent }],
+    },
 ];
```
Comment (Karhalll) [inline src/app/app-routing.module.ts:12]: There is no path for registration that is already in master. But i guess it will be there after resolving conflicts.

File: src/app/app.component.ts [hunk 2]
```diff
@@ -12,11 +13,14 @@ export class AppComponent {
...
-        this._gateway.fetchPosts()
-            .subscribe(response => {
-                this._postService.setPosts(response.posts);
-            })
+        this._gateway.fetchPosts().subscribe((response) => {
+            this._postService.setPosts(response);
+        });
+        this._gateway.fetchSubreddits().subscribe((response) => {
+            this._subredditService.setSubreddits(response);
+        });
```
Comment (Karhalll) [inline src/app/app.component.ts:24]: Is it neccessary to have this here? That mean that we are fetching posts and subreddits every time even if there are no components to show them. Like for landing page/registration form/ (future) login form.

File: src/app/components/authorized/welcome-page/welcome-page.component.html [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+            </div>
+
+            <div class="registrationContainer">
+                <h2>If you do not have account, please register.</h2>
+                <button class="registrationBtn" type="submit">Sign in</button>
+            </div>
+        </div>
+    </div>
+
+</div>
```
Comment (Karhalll) [inline src/app/components/authorized/welcome-page/welcome-page.component.html:23]: Thanks for that :)

File: src/app/components/authorized/welcome-page/welcome-page.component.scss [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      @include button;
+      justify-self: center;
+    };
+    display: grid;
+    grid-template-columns: 4fr 1fr;
+    column-gap: 1rem;
+    grid-template-areas: 
+      "sidePanelMessage subBtn";
+  };
+}
```
Comment (Karhalll) [inline src/app/components/authorized/welcome-page/welcome-page.component.scss:76]: I think some of that should be at the app-root component. Since landing page is really just for new useres. (navigationBar for exapmle)

File: src/app/components/authorized/feed/feed.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    id: number;
+    title: string;
+    date_created: Date;
+    subreddit: string;
+    author: string;
+    commentCount: number;
+    post_type: string;
+    posted_url?: string;
+    description: string;
+}
```
Comment (Karhalll) [inline src/app/components/authorized/feed/feed.component.spec.ts:17]: Don't you have that interface created as type somewhere in the app already?

File: src/assets/posts.json [hunk 6]
```diff
@@ -1,25 +1,23 @@
...
+        "title": "Second post",
+        "date_created": "2020-11-30T17:10:56.782Z",
+        "subreddit": "Subreddit2",
+        "author": "Author2",
+        "commentCount": 3,
+        "post_type": "url",
+        "posted_url": "https://seznam.cz",
+        "description": "Post about anything"
+    }
+]
```
Comment (Karhalll) [inline src/assets/posts.json:23]: Do you need this? Since you are mocking posts in tests and feed is functional on backend.

File: src/assets/subreddits.json [hunk 7]
```diff
@@ -1,20 +1,18 @@
...
+    },
+    {
+        "name": "Subreddit2",
+        "title": "Subreddit2 title",
+        "date_created": "2020-11-30T17:10:56.782Z",
+        "description": "Subreddit2 obout anything",
+        "userCount": 20,
+        "author": "Author2"
+    }
+]
```
Comment (Karhalll) [inline src/assets/subreddits.json:18]: Same question as in posts.json

**Review by Karhalll — APPROVED** [review]
You have unresolved conflicts! Other then that it looks ok for me. I'am little bit confused if you have post and subreddit service why are we fatching those in gateaway service. But that is just me not realy being on front end so far.

File: src/app/app-routing.module.ts (hunk 1 above)
Comment (AndyM22) [reply src/app/app-routing.module.ts]: Yes.

File: src/app/app.component.ts (hunk 2 above)
Comment (AndyM22) [reply src/app/app.component.ts]: No, it was before authorizedComponent was created and I forgot to solve this problem. Now it's in the AuthorizedComponent

File: src/app/components/authorized/welcome-page/welcome-page.component.html (hunk 3 above)
Comment (AndyM22) [reply src/app/components/authorized/welcome-page/welcome-page.component.html]: no problem :)

File: src/app/components/authorized/welcome-page/welcome-page.component.scss (hunk 4 above)
Comment (AndyM22) [reply src/app/components/authorized/welcome-page/welcome-page.component.scss]: No, the same problem as before, forgot to solve this problem after creating authorizedComponent.

File: src/app/components/authorized/feed/feed.component.spec.ts (hunk 5 above)
Comment (AndyM22) [reply src/app/components/authorized/feed/feed.component.spec.ts]: Yes.

File: src/assets/posts.json (hunk 6 above)
Comment (AndyM22) [reply src/assets/posts.json:23]: If it wouldn't mind you I would keep it since it helps me to develop frontend faster (because of my slow computer).

File: debug.log [hunk 8]
```diff
@@ -15,3 +15,40 @@
 [1130/214333.166:ERROR:directory_reader_win.cc(43)] FindFirstFile: Syst�m nem�e nal�zt uvedenou cestu. (0x3)
```
Comment (SimProch) [inline debug.log:15]: This file should be in git ignore.

File: src/app/app.component.spec.ts [hunk 9]
```diff
@@ -1,18 +1,12 @@
...
-import { PostService } from './services/post.service';
-import { SubredditService } from './services/subreddit.service';
 
 describe('AppComponent', () => {
     beforeEach(async () => {
         await TestBed.configureTestingModule({
-            imports: [BrowserModule, AppRoutingModule, HttpClientModule],
-            providers: [PostService, SubredditService, GatewayService],
             declarations: [AppComponent],
+            schemas: [CUSTOM_ELEMENTS_SCHEMA],
```
Comment (SimProch) [inline src/app/app.component.spec.ts:9]: Why is this necessary in here?

File: src/app/app.component.ts [hunk 10]
```diff
@@ -10,9 +8,5 @@ import { PostService } from './services/post.service';
 export class AppComponent {
     title = 'IceAge-Reddit-Frontend';
 
-    constructor(private _gateway: GatewayService, private _postService: PostService) {
-        this._gateway.fetchPosts().subscribe((response) => {
-            this._postService.setPosts(response.posts);
-        });
-    }
+    constructor() {}
```
Comment (SimProch) [inline src/app/app.component.ts:11]: Empty constructor is useless here.

File: src/app/components/authorized/post/post.component.spec.ts [hunk 11]
```diff
@@ -58,6 +58,31 @@ describe('PostComponent', () => {
         expect(component).toBeTruthy();
     });
 
+    it('postId should be 1', () => {
```
Comment (SimProch) [inline src/app/components/authorized/post/post.component.spec.ts:61]: This entire test suite is pretty much useless. You are providing mocked data and you are testing if the data is really mocked. There is no logical testing in here.

File: src/app/components/authorized/subreddit/subreddit.component.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    selector: 'app-subreddit',
+    templateUrl: './subreddit.component.html',
+    styleUrls: ['./subreddit.component.scss'],
+})
+export class SubredditComponent implements OnInit {
+    @Input() subredditName: string;
+
+    subreddit: Subreddit;
+
+    constructor(private _subredditService: SubredditService) {}
```
Comment (SimProch) [inline src/app/components/authorized/subreddit/subreddit.component.ts:15]: Useless constructor

File: src/app/components/authorized/subreddit/subreddit.component.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+    ngOnInit(): void {
+        // this.subreddit = this._subredditService.getSubreddit(this.subredditName);
```
Comment (SimProch) [inline src/app/components/authorized/subreddit/subreddit.component.ts:18]: Commented code should not be part of code review as it will almost always not pass

File: src/app/components/authorized/subreddits/subreddits.component.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+})
+export class SubredditsComponent {
+    public _subreddits: Subreddit[];
+
+    constructor(private _subredditService: SubredditService) {
+        this._subredditService.subreddits$.subscribe(
+            (subreddits) => (this._subreddits = subreddits),
+        );
+    }
+    // onSelect(subreddit) {
```
Comment (SimProch) [inline src/app/components/authorized/subreddits/subreddits.component.ts:18]: Same as previously mentioned - commented code should not be part of CR.

File: src/app/components/welcome-page/welcome-page.component.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+    selector: 'app-welcome-page',
+    templateUrl: './welcome-page.component.html',
+    styleUrls: ['./welcome-page.component.scss'],
+})
+export class WelcomePageComponent implements OnInit {
+    constructor() {}
```
Comment (SimProch) [inline src/app/components/welcome-page/welcome-page.component.ts:9]: Useless ngOnInit and constructor.

File: src/app/services/gateway.service.ts [hunk 16]
```diff
@@ -1,25 +1,42 @@
...
 
 @Injectable({
     providedIn: 'root',
 })
 export class GatewayService {
-    private _data = 'assets/posts.json';
+    private _mockPostData = 'assets/posts.json';
+    private _mockSubredditData = 'assets/subreddits.json';
+
+    private _baseUrl = 'http://localhost:3000/api/v1/';
```
Comment (SimProch) [inline src/app/services/gateway.service.ts:15]: The comments in this file are not saying much. 
```
// subreddits
function fetchSubreddits() {...}
```

You make a comment where the function itself already mentions it. It should not be in there.

Furthermore, the API names are odd. You can't really use base url like this, especially because it will be different for different environments. This may already be solved with heroku deployment. 

Similar to previous comment, `_feedData` and `baseUrl/feed` are telling exactly the same. It does not have to be a property.

File: src/app/services/post.service.spec.ts [hunk 17]
```diff
@@ -1,16 +1,65 @@
...
+
+    it('setPosts function should add postsData to testPosts', () => {
+        service.setPosts(postsData);
+        expect(service.posts$).toEqual(new BehaviorSubject(postsData));
+    });
 
-  beforeEach(() => {
-    TestBed.configureTestingModule({
+    it('getAllPosts should return null', () => {
+        expect(service.getAllPosts()).toEqual(new BehaviorSubject(null));
```
Comment (SimProch) [inline src/app/services/post.service.spec.ts:52]: This test is incorrect. `new BehaviorSubject(null)` does not return null, rather a behavior subject that contains the value `null`. Tests are specifications and must be descriptive and true.

File: src/assets/posts.json (hunk 6 above)
Comment (SimProch) [reply src/assets/posts.json:23]: You should not use it like this, though. Instead, you should use only one. It makes sense in some big data cases that you are using mocked data for development rather than real data as well. 

Think about using mocked data on your machine only and doing it sustainably so you don't need to change that every time.

**Review by SimProch — CHANGES_REQUESTED** [review]
See comments and think about changes. If you need some more information, contact me.

File: debug.log (hunk 8 above)
Comment (AndyM22) [reply debug.log:15]: It is I don't know why it was pushed.

File: src/app/app.component.spec.ts (hunk 9 above)
Comment (AndyM22) [reply src/app/app.component.spec.ts:9]: It is not I already deleted and fixed problem with adding missing components.

File: src/app/app.component.ts (hunk 10 above)
Comment (AndyM22) [reply src/app/app.component.ts:11]: Deleted.

File: src/app/components/welcome-page/welcome-page.component.ts (hunk 15 above)
Comment (AndyM22) [reply src/app/components/welcome-page/welcome-page.component.ts:9]: Deleted.

File: src/app/services/gateway.service.ts (hunk 16 above)
Comment (AndyM22) [reply src/app/services/gateway.service.ts:13]: Ok, I have deleted the properties and comments. The baseURL stayed the same, because Martin will change it with heroku deployment.

File: src/app/components/authorized/subreddit/subreddit.component.ts (hunk 13 above)
Comment (AndyM22) [reply src/app/components/authorized/subreddit/subreddit.component.ts]: I deleted the content of subreddit and subreddits components since I haven't solved their correct functionality yet and it's not the goal of this ticket. I will fix it in the ticket where it's supposed to be.

File: src/assets/posts.json (hunk 6 above)
Comment (AndyM22) [reply src/assets/posts.json]: I deleted both files. I haven't figured out how to use mocked data only on my machine, but it works with with backend acceptably.

File: src/app/components/authorized/post/post.component.spec.ts (hunk 11 above)
Comment (AndyM22) [reply src/app/components/authorized/post/post.component.spec.ts]: I deleted unnecessary code and only left tests to check if getPost method was called from service. Is that alright??

File: src/app/app-routing.module.ts [hunk 18]
```diff
@@ -1,19 +1,24 @@
...
+    { path: 'subreddits', component: SubredditsComponent },
+    { path: 'registration', component: RegistrationFormComponent },
 ];
 
 @NgModule({
     imports: [RouterModule.forRoot(routes)],
     exports: [RouterModule],
 })
 export class AppRoutingModule {}
+export const routingComponents = [
```
Comment (SimProch) [inline src/app/app-routing.module.ts:19]: Why is this constant in here, what is the purpose of it?

File: src/app/components/authorized/post/post.component.spec.ts [hunk 19]
```diff
@@ -45,39 +44,16 @@ describe('PostComponent', () => {
...
-    it('Should display post subreddit', () => {
-        expect(aPostSubreddit.textContent).toContain(component.postSubreddit);
-    });
-
-    it('Should display post commentCount', () => {
-        expect(divPostCommentCount.textContent).toContain(component.postCommentCount.toString());
+    it('should make a call to postService.getPost(postId)', () => {
+        spyOn(component['_postService'], 'getPost').and.callThrough();
+        component.ngOnInit();
+        expect(component['_postService'].getPost).toHaveBeenCalledWith(component.postId);
```
Comment (SimProch) [inline src/app/components/authorized/post/post.component.spec.ts:57]: Nice usage of spies. I'd suggest making more tests in here, though.

File: src/types/posts.ts [hunk 20]
```diff
@@ -3,6 +3,7 @@ export interface PostResponse {
 }
 
 export interface Post {
+    [x: string]: any;
```
Comment (SimProch) [inline src/types/posts.ts:6]: This doesn't sound right. Why is this necessary?

This type of typing makes sense when you have vague information of structure. For example, we have a page that requires a user to sign in. For some users, an endpoint returns an object that has 10 items. For others, it has 20. For admin, it could go all the way to 200. You don't want to write all the possibilities (or you might, depends on situation).

Posts in our subreddit are specific. They have a specific list of properties - at least according to apiary. So why is this necessary?

**Review by SimProch — CHANGES_REQUESTED** [review]

File: src/types/posts.ts (hunk 20 above)
Comment (AndyM22) [reply src/types/posts.ts]: It was added by accident when I was choosing from the options offered by ctrl+.

File: src/app/app-routing.module.ts (hunk 18 above)
Comment (AndyM22) [reply src/app/app-routing.module.ts:19]: Cosmetic experiment from one video...
It has no usage, now it's gone

**Review by SimProch — APPROVED** [review]
